pkg/storegateway/postings_codec.go

```diff
@@ -97,6 +105,141 @@ func diffVarintSnappyDecode(input []byte) (index.Postings, error) {
+	}
+
+	// Compressed postings
+	compressedPostings := snappy.Encode(result[codecLen+matchersWritten:], varintPostings)
+
+	result = result[:codecLen+matchersWritten+len(compressedPostings)]
+	return result, nil
+}
+
+// encodeMatchers needs to be called with the precomputed length of the encoded matchers from encodedMatchersLen
+func encodeMatchers(totalLen int, matchers []*labels.Matcher, dest []byte) (written int, _ error) {
+	if len(dest) < totalLen {
+		return 0, fmt.Errorf("too small buffer to encode matchers: need at least %d, got %d", totalLen, dest)
+	}
+	written += binary.PutUvarint(dest, uint64(len(matchers)))
+	for _, m := range matchers {
+		written += binary.PutUvarint(dest[written:], uint64(len(m.Name)))
+		written += copy(dest[written:], m.Name)
+
+		dest[written] = byte(m.Type)
+		written++
+
+		written += binary.PutUvarint(dest[written:], uint64(len(m.Value)))
+		written += copy(dest[written:], m.Value)
+	}
+	return written, nil
+}
+
+func encodedMatchersLen(matchers []*labels.Matcher) int {
+	matchersLen := varint.UvarintSize(uint64(len(matchers)))
+	for _, m := range matchers {
+		matchersLen += varint.UvarintSize(uint64(len(m.Name)))
+		matchersLen += len(m.Name)
+		matchersLen++ // 1 byte for the type
+		matchersLen += varint.UvarintSize(uint64(len(m.Value)))
+		matchersLen += len(m.Value)
+	}
+	return matchersLen
+}
+
+// diffVarintSnappyMatchersDecode retains pointers to the buffer.
+func diffVarintSnappyMatchersDecode(input []byte) (index.Postings, []*labels.Matcher, error) {
+	if !isDiffVarintSnappyWithmatchersEncodedPostings(input) {
+		return nil, nil, errors.New("dm header not found")
+	}
+
+	codecLen := len(codecHeaderSnappyWithmatchers)
+	matchers, matchersLen, err := decodeMatchers(input[codecLen:])
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "decoding matchers")
+	}
+	raw, err := snappy.Decode(nil, input[codecLen+matchersLen:])
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "snappy decode")
+	}
+
+	return newDiffVarintPostings(raw), matchers, nil
+}
+
+// decodeMatchers retains points to the buffer
+func decodeMatchers(src []byte) ([]*labels.Matcher, int, error) {
+	initialLength := len(src)
+	numMatchers, numMatchersLen := varint.Uvarint(src)
+	src = src[numMatchersLen:]
+	if numMatchers == 0 {
+		return nil, numMatchersLen, nil
+	}
+	matchers := make([]*labels.Matcher, 0, numMatchers)
+
+	var (
+		typ       labels.MatchType
+		labelName string
+		value     string
+	)
+	for i := uint64(0); i < numMatchers; i++ {
+		n, nLen := varint.Uvarint(src)
+		src = src[nLen:]
+		labelName = yoloString(src[:n])
+		src = src[n:]
+
+		typ = labels.MatchType(src[0])
+		src = src[1:]
+
+		n, nLen = varint.Uvarint(src)
+		src = src[nLen:]
+		value = yoloString(src[:n])
+		src = src[n:]
+
+		// TODO dimitarvdimitrov add a test to make sure the matcher is compiled (maybe we can even take it from the compiled query matchers?)
```

Comment:
I'm not sure what you want to test here :)

Reply:
because i initially created a `labels.Matcher` struct and assigned each field instead of using `NewMatcher`. So that one would have paniced when you try to use it for matching.


But that panic would be caught in the following PRs when we actually have some matchers. I'm ok with writing fewer tests, I'll remove the TODO :)